### ISSUE TITLE
[Hardware Quirk] GPU hangs if one shader program has a geo component and the other doesn't

### ISSUE THREAD

**Issue by TheGag96** [fincs/citro3d#56]
In working [this demo](https://github.com/TheGag96/3ds-scanline-offset-shader/tree/18d7991f529675a008dc82cadc93f32b8fcbccaa) using two different shader programs to draw a scene, I was getting a hard GPU hang on real hardware (and not on Citra). I narrowed the cause down to one shader I was using having a geo part as well as a vertex part and the other one just having a vertex part. If you flip-flopped which shader you used every frame, there'd be no issue.

Here's a [demonstration based on the composite_scene example](https://gist.github.com/TheGag96/0b844e7fc7109df54a399beb3182f76f). This exhibits the same behavior I was running into. However, I just went and made [another one based on the lenny example](https://gist.github.com/TheGag96/033c582ba37cc5a691aaa3da061b7d0c) and this one doesn't seem to hang. Is something done in citro2d the culprit? Or is the lenny-based example not hanging because the shader programs are basically the same aside from the geo part?

I was told to submit an issue here. Hopefully this helps out!

**Comment by fordcars** [comment]
I know I'm really late to the game, but I ran into this exact same issue. Turns out the culprit on my end were uniforms; if you switch from a shader with a geo component to a shader without one, and then back to the geo shader, all uniforms are cleared. Resetting my uniform values after switching shaders fixed this issue for me.

**Comment by TheGag96** [comment]
Yeah, I ended up doing the same thing... And wow, boosted performance? How's that?

**Comment by fordcars** [comment]
Looks like switching between shaders with and without geo components really put a strain on my system. The same code with the dummy geo shader for triangles had a big improvement in performance, but that might be because of the *insane* amount of shader changes I am doing!

**Comment by Master-gtaAndroid** [comment]
Good afternoon friends, can you help me as soon as I start with this 3ds thing, what happens is that when I do make it generates an error, I already installed everything but it tells me.


make[1]: se entra en el directorio '/home/jv866/Downloads/FTP-GMX-master/FTP-GMX-master'
linking FTP-GMX-2.2.elf
c:/devkitpro/devkitarm/bin/../lib/gcc/arm-none-eabi/12.1.0/../../../../arm-none-eabi/bin/ld.exe: cannot find -lsf2d: No such file or directory
collect2.exe: error: ld returned 1 exit status
make[2]: *** [/opt/devkitpro/devkitARM/3ds_rules:42: /home/jv866/Downloads/FTP-GMX-master/FTP-GMX-master/FTP-GMX-2.2.elf] Error 1
make[1]: *** [Makefile.3ds:134: build] Error 2
make[1]: se sale del directorio '/home/jv866/Downloads/FTP-GMX-master/FTP-GMX-master'
